lobby/src/main/java/org/triplea/lobby/server/db/UserController.java

```diff
@@ -109,24 +113,23 @@ public void createUser(
 
   @Override
   public boolean login(final String username, final HashedPassword hashedPassword) {
-    try (Connection con = connection.get()) {
-      final HashedPassword actualPassword = getPassword(username);
-      if (actualPassword == null) {
-        return false;
-      }
-      Preconditions.checkState(actualPassword.isBcrypted());
-      if (!BCrypt.checkpw(hashedPassword.value, actualPassword.value)) {
-        return false;
-      }
-      // update last login time
-      try (PreparedStatement ps =
-          con.prepareStatement("update lobby_user set last_login = ? where username = ?")) {
-        ps.setTimestamp(1, Timestamp.from(Instant.now()));
-        ps.setString(2, username);
-        ps.execute();
-        con.commit();
-        return true;
-      }
+    final HashedPassword actualPassword = getPassword(username);
+    if (actualPassword == null) {
+      return false;
+    }
+    Preconditions.checkState(actualPassword.isBcrypted());
+    if (!BCrypt.checkpw(hashedPassword.value, actualPassword.value)) {
+      return false;
+    }
+    // update last login time
```

Comment:
This is a place where it's a shame we have to read password. We could do password validation and update in one go with a where clause:

```
update lobby_user set last_login = now() where username = ? and password = ?
```

Benefits there is the update is fully atomic, and the encrypted password never leaves the database. Having functions to read passwords from DB is not the best. If an attacker is able to call arbitrary server commands somehow, that could be used to read all encrypted passwords from DB, while on the other hand if it is one-way only then the best they could do is a brute force attack against unknown values.

Reply:
I agree that having to read a password before comparing it isn't optimal, but it's recommended security practice to store passwords this way.

Reply:
Bcrypt is a good 'slow' hashing mechanism; I wonder if there is another API we can call to compute the expected hash and use that. Essentially avoid the comparison of using `Bcrypt.checkPw`, IMO it's more the APIs we are using and less the encryption mechanism.

Reply:
I mean if you insist on comparing hashes directly, you could of course query the db, to give you the metadata information of the hash (amount of rounds, salt) and hash the "original password" using this information.
This will result in the same hash that can be compared with the db.

However if you're looking for a way to compare passwords without any previous query whatsoever, there isn't any possibility because the whole point of hashing passwords with salt is that they need to be compared individually so an attacker can't pre-compute all the hashes

Reply:
Interesting, thanks for the respons @RoiEXLab ; something I'll need to think about.